README.md
This is the README file for MTK-Cmd,
MTK CLI infrastructure and basic commands.

## Description

MTK-Cmd provides the MTK CLI
infrastructure and basic commands

## Installation

This package uses Dist::Zilla.

Use

dzil build

to create a release tarball which can be
unpacked and installed like any other EUMM
distribution.

perl Makefile.PL

make

make test

make install

## Documentation

Please see perldoc MTK::Cmd.



Changes
Revision history for MTK-Cmd

0.16	2012-12-26
        First public version, released on an unsuspecting world.


lib/MTK/Cmd.pm
package MTK::Cmd;
# ABSTRACT: MTK CLI infrastructure and basic commands

use 5.010_000;
use mro 'c3';
use feature ':5.10';

use Moose;
use namespace::autoclean;

# use IO::Handle;
# use autodie;
# use MooseX::Params::Validate;
# use Carp;
# use English qw( -no_match_vars );
# use Try::Tiny;

# extends ...
extends 'MooseX::App::Cmd';
# has ...
# with ...
# initializers ...

# your code here ...

no Moose;
__PACKAGE__->meta->make_immutable;

1;

__END__

=head1 NAME

MTK::Cmd - Command base class.

=head1 SYNOPSIS

    use MTK::Cmd;
    my $Cmd = MTK::Cmd::->new();
    $Cmd->run();

=head1 DESCRIPTION

This class is the base class for any command implemented by its subclasses.

It is a mere requirement by App::Cmd. Don't mess with it.

=head1 ACKNOWLEDGEMENT

This module was originally developed for eGENTIC Systems. With approval from eGENTIC Systems,
this module was generalized and published, for which the authors would like to express their
gratitude.

=cut


lib/MTK/Cmd/Command.pm
package MTK::Cmd::Command;
# ABSTRACT: baseclass for any MTK CLI command

use 5.010_000;
use mro 'c3';
use feature ':5.10';

use Moose;
use namespace::autoclean;

# use IO::Handle;
# use autodie;
# use MooseX::Params::Validate;
# use Carp;
# use English qw( -no_match_vars );
# use Try::Tiny;
use Config::Yak;
use Log::Tree;

# extends ...
extends 'MooseX::App::Cmd::Command';
# has ...
has '_config' => (
    'is'    => 'rw',
    'isa'   => 'Config::Yak',
    'lazy'  => 1,
    'builder' => '_init_config',
    'accessor' => 'config',
);

has '_logger' => (
    'is'    => 'rw',
    'isa'   => 'Log::Tree',
    'lazy'  => 1,
    'builder' => '_init_logger',
    'accessor' => 'logger',
);
# with ...
# initializers ...
sub _init_config {
    my $self = shift;

    my $Config = Config::Yak::->new({
        'locations'     => $self->_config_locations(),
    });

    return $Config;
}

sub _init_logger {
    my $self = shift;

    my $Logger = Log::Tree::->new($self->_log_facility());

    return $Logger;
}

# your code here ...
sub _config_locations {
    return [qw(conf /etc/mtk)];
}

sub _log_facility {
    return 'mtk-cli';
}

sub ask_yesno {
    my $self = shift;
    my $msg = shift;
    print $msg. ' [y/N]: ';
    ## no critic (ProhibitExplicitStdin)
    my $resp = <STDIN>;
    ## use critic
    chomp($resp);
    if ( $resp =~ m/(1|Yes|Ja|Y)/i ) {
        return 1;
    }
    return;
}

sub ask_yesnoall {
    my $self = shift;
    my $msg = shift;
    print $msg. ' [y/N/a]: ';
    ## no critic (ProhibitExplicitStdin)
    my $resp = <STDIN>;
    ## use critic
    chomp($resp);
    if ( $resp =~ m/(1|Yes|Ja|Y)/i ) {
        return 1;
    }
    elsif ( $resp =~ m/(a|All|alle)/i ) {
        return 2;
    }
    return;
}

sub ask_select {
    my ( $self, $msg, @options ) = @_;

    # let user select on of the options provided
    while (1) {
        print $msg. "\n";
        my $i = 0;
        foreach my $opt (@options) {
            print "[$i] $opt\n";
            $i++;
        }
        my $num = ask_number( 'Print enter any number between 0 and ' . $i . '. Press enter to abort' );
        if ( defined($num) && $options[$num] ) {
            return $options[$num];
        }
        else {
            return;
        }
    }

    return;
}

sub ask_number {
    my $self = shift;
    my $msg = shift;
    print $msg. ': ';
    ## no critic (ProhibitExplicitStdin)
    my $resp = <STDIN>;
    ## use critic
    chomp($resp);
    if ( $resp =~ m/^\s*(\d+)\s*$/ ) {
        return $1;
    }
    return;
}

sub ask_string {
    my $self = shift;
    my $msg = shift;
    print $msg. ': ';
    ## no critic (ProhibitExplicitStdin)
    my $resp = <STDIN>;
    ## use critic
    chomp($resp);
    return $resp;
}

no Moose;
__PACKAGE__->meta->make_immutable;

1;

__END__

=head1 NAME

MTK::Cmd::Command - baseclass for any MTK CLI command

=head1 DESCRIPTION

Base class for any MTK CLI command.

=method ask_number

Helper to ask for a number on STDIN.

=method ask_select

Helper to ask for the selection of a given set of options from STDIN.

=method ask_string

Helper to ask for a string on STDIN.

=method ask_yesno

Helper to ask for a true/false on STDIN.

=method ask_yesnoall

Helper to ask for a true/false/all on STDIN.

=head1 ACKNOWLEDGEMENT

This module was originally developed for eGENTIC Systems. With approval from eGENTIC Systems,
this module was generalized and published, for which the authors would like to express their
gratitude.

=cut


lib/MTK/Cmd/Command/delay.pm
package MTK::Cmd::Command::delay;
# ABSTRACT: keep a mysql slave a given number of seconds behind the master

use 5.010_000;
use mro 'c3';
use feature ':5.10';

use Moose;
use namespace::autoclean;

# use IO::Handle;
# use autodie;
# use MooseX::Params::Validate;
# use Carp;
# use English qw( -no_match_vars );
# use Try::Tiny;
use MTK::DB;
use MTK::DB::Credentials;
use Linux::Pidfile;

# extends ...
extends 'MTK::Cmd::Command';
# has ...
has '_pidfile' => (
    'is'    => 'ro',
    'isa'   => 'Linux::Pidfile',
    'lazy'  => 1,
    'builder' => '_init_pidfile',
);
# with ...
# initializers ...
sub _init_pidfile {
    my $self = shift;

    my $Pidfile = Linux::Pidfile::->new({
        'logger' => $self->logger(),
        'pidfile' => $self->config()->get( 'MTK::Delay::Pidfile', { Default => '/var/run/mtk-delay.pid', } ),
    });

    return $Pidfile;
}

# your code here ...
sub execute {
    my $self = shift;

    $self->_pidfile()->create() or die('Script already running.');

    my $running = 1;

    my $host     = $self->config()->get('MTK::Delay::DB::Host') || 'localhost';
    my $port     = $self->config()->get('MTK::Delay::DB::Port') || 3306;
    my $database = $self->config()->get('MTK::Delay::DB::DBName')    or croak('Need MTK::Delay::DB::DBName');
    my $table    = $self->config()->get('MTK::Delay::DB::TableName') or croak('Need MTK::Delay::DB::TableName');
    my $Creds    = MTK::DB::Credentials::->new(
        {
            'config'   => $self->config(),
            'hostname' => $host,
            'keys'     => [qw(MTK::Delay::DB)],
            'logger'   => $self->logger(),
        }
    );
    my $username = $Creds->username();
    my $password = $Creds->password();
    my $DB       = $Creds->dbh();
    $Creds = undef;

    if ( !$DB ) {
        $self->logger()->log( message => 'Database not available. User: '.$username.', Host: '.$host.', Port: '.$port.', Error: ' . DBI->errstr, level => 'emerg', );
        return;
    }

    my $query = 'SELECT COUNT(*) FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?';
    my $sth   = $DB->prepexec( $query, $database, $table );
    my $count = $sth->fetchrow_array($sth);
    $sth->finish();
    my $sss;
    if ( $count > 0 ) {
        $query = 'SELECT ts FROM `'.$database.'`.`'.$table.'` WHERE id=0 LIMIT 1';
        $sss   = $DB->prepare($query);
    }

    my $holdbacktime = $self->config()->get('MTK::Delay::Holdbacktime');
    my $checktime    = $self->config()->get('MTK::Delay::Checktime');

    while ($running) {

        # reconnect if necessary

        if ( !$DB->check_connection() ) {
            $self->logger()->log( message => 'Could not connect to DSN: ' . $DB->dsn() . ' - ' . $DB->errstr(), level => 'alert', );
        }

        # Record the slave status
        my $secbehind = -1;
        if ($sss) {
            $sss->execute();
            $secbehind = $sss->fetchrow_array();
        }
        else {
            $secbehind = $DB->get_sec_behind_master();
        }

        $self->logger()->log( message => "Seconds_behind_Master: $secbehind", level => 'debug', );

        if ( $secbehind < $holdbacktime ) {
            $DB->lock_tables( { 'heartbeat' => 'WRITE', } );
            $self->logger()->log( message => "Locking $table: $query", level => 'debug', );
        }
        else {
            $DB->unlock_tables();
            $self->logger()->log( message => "Unlocking $table", level => 'debug', );
        }
        sleep($checktime);
    }
    $sss->finish() if $sss;
    $DB->disconnect();
    return 1;
}

sub abstract {
    return 'Keep an MySQL Slave a given number of seconds behind the master';
}

sub DEMOLISH {
    my $self = shift;

    $self->_pidfile()->remove();

    return 1;
}

no Moose;
__PACKAGE__->meta->make_immutable;

1;

__END__

=head1 NAME

MTK::Cmd::Command::delay - keep a mysql slave a given number of seconds behind the master

=method DEMOLISH

Remove our pidfile.

=method abstract

Workaround.

=method execute

Run the delay command.

=head1 ACKNOWLEDGEMENT

This module was originally developed for eGENTIC Systems. With approval from eGENTIC Systems,
this module was generalized and published, for which the authors would like to express their
gratitude.

=cut


lib/MTK/Cmd/Command/heartbeat.pm
package MTK::Cmd::Command::heartbeat;
# ABSTRACT: write a timestamp to a db, every second

use 5.010_000;
use mro 'c3';
use feature ':5.10';

use Moose;
use namespace::autoclean;

# use IO::Handle;
# use autodie;
# use MooseX::Params::Validate;
# use Carp;
use English qw( -no_match_vars );
# use Try::Tiny;
use Linux::Pidfile;

use MTK::DB;
use MTK::DB::Credentials;

# extends ...
extends 'MTK::Cmd::Command';
# has ...
has '_pidfile' => (
    'is'    => 'ro',
    'isa'   => 'Linux::Pidfile',
    'lazy'  => 1,
    'builder' => '_init_pidfile',
);
# with ...
# initializers ...
sub _init_pidfile {
    my $self = shift;

    my $Pidfile = Linux::Pidfile::->new({
        'logger' => $self->logger(),
        'pidfile' => $self->config()->get( 'MTK::Heartbeat::Pidfile', { Default => '/var/run/mtk-heartbeat.pid', } ),
    });

    return $Pidfile;
}

# your code here ...
sub execute {
    my $self = shift;

    $self->_pidfile()->create() or die('Script already running.');

    my $hostname = $self->config()->get('MTK::Heartbeat::DB::Hostname')  || 'localhost';
    my $port     = $self->config()->get('MTK::Heartbeat::DB::Port')      || 3306;
    my $db       = $self->config()->get('MTK::Heartbeat::DB::DBName')    || 'REPLICHECK';
    my $table    = $self->config()->get('MTK::Heartbeat::DB::TableName') || 'heartbeat';

    my $Creds = MTK::DB::Credentials::->new(
        {
            'config'   => $self->config(),
            'hostname' => $hostname,
            'keys'     => [qw(MTK::Heartbeat::DB)],
            'logger'   => $self->logger(),
        }
    );
    my $DB  = $Creds->dbh();
    my $dsn = $DB->dsn();
    $Creds = undef;

    if ( !$DB || !$DB->valid() ) {
        $self->logger()->log( message => "Database not available. DSN: $dsn, Error: " . DBI->errstr, level => 'error', );
        return;
    }

    my $query = 0;    # Hold query strings
    my $prepq = 0;    # Hold prepared query

    $query = 'UPDATE ' . $table . ' SET ts=? WHERE id=0';
    $self->logger()->log( message => 'Query: '.$query, level => 'debug', );
    $prepq = $DB->prepare($query);

    my $running = 1;
    while ($running) {
        my $current_time = time();
        $self->logger()->log( message => 'Updating to current time: '.$current_time, level => 'debug', );

        # use an alarm timer to prevent hanging
        # reconnect if connection dies
        my $timeout      = 30;
        my $prev_timeout = 0;
        my $ok           = eval {
            local $SIG{ALRM} = sub { die "alarm-mtk-hearbeat\n"; };
            $prev_timeout = alarm $timeout;
            if ( !$prepq->execute($current_time) ) {

                # reconnect if necessary
                $DB->check_connection();
            }

            # restore previous alarm, if any
            alarm $prev_timeout;
            1;
        };

        # make sure the alarm is off
        alarm $prev_timeout;
        if ( $EVAL_ERROR && $EVAL_ERROR eq "alarm-mtk-hearbeat\n" ) {
            $self->logger()->log( message => 'Connection timed out after '.$timeout, level => 'warning', );
            sleep 30;
        }
        elsif ( !$ok ) {
            $self->logger()->log( message => 'Eval failed somehow (unknown error).', level => 'warning', );
        }
        sleep(1);
    }
    $prepq->finish();
    $DB->disconnect();
    return 1;
}

sub abstract {
    return 'Constantly write a timestamp to a given MySQL table. Used as a heartbeat for mtk delay.';
}

sub DEMOLISH {
    my $self = shift;

    $self->_pidfile()->remove();

    return 1;
}

no Moose;
__PACKAGE__->meta->make_immutable;

1;

__END__

=head1 NAME

MTK::Cmd::Command::heartbeat - write a timestamp to a db, every second

=method DEMOLISH

Remove our pidfile.

=method abstract

Workaround.

=method execute

run the heartbeat command.

=head1 ACKNOWLEDGEMENT

This module was originally developed for eGENTIC Systems. With approval from eGENTIC Systems,
this module was generalized and published, for which the authors would like to express their
gratitude.

=cut


lib/MTK/Cmd/Command/doall.pm
package MTK::Cmd::Command::doall;
# ABSTRACT: execute a given SQL stmt on any db and/or table

use 5.010_000;
use mro 'c3';
use feature ':5.10';

use Moose;
use namespace::autoclean;

# use IO::Handle;
# use autodie;
# use MooseX::Params::Validate;
# use Carp;
use English qw( -no_match_vars );
# use Try::Tiny;
use File::Basename qw();

use MTK::DB;

# extends ...
extends 'MTK::Cmd::Command';

# has ...
has 'dbs' => (
    'is'            => 'ro',
    'isa'           => 'Str',
    'required'      => 0,
    'traits'        => [qw(Getopt)],
    'cmd_aliases'   => 'd',
    'documentation' => 'Process these databases',
);

has 'excludes' => (
    'is'            => 'ro',
    'isa'           => 'Str',
    'required'      => 0,
    'traits'        => [qw(Getopt)],
    'cmd_aliases'   => 'e',
    'documentation' => 'Excludes thoses patterns',
);

has 'tables' => (
    'is'            => 'ro',
    'isa'           => 'Str',
    'required'      => 0,
    'traits'        => [qw(Getopt)],
    'cmd_aliases'   => 't',
    'documentation' => 'Process all tables',
);

has 'user' => (
    'is'            => 'ro',
    'isa'           => 'Str',
    'required'      => 1,
    'traits'        => [qw(Getopt)],
    'cmd_aliases'   => 'u',
    'documentation' => 'Login',
);

has 'pass' => (
    'is'            => 'ro',
    'isa'           => 'Str',
    'required'      => 1,
    'traits'        => [qw(Getopt)],
    'cmd_aliases'   => 'p',
    'documentation' => 'Password',
);

has 'port' => (
    'is'            => 'ro',
    'isa'           => 'Int',
    'required'      => 0,
    'default'       => 3306,
    'traits'        => [qw(Getopt)],
    'cmd_aliases'   => 'o',
    'documentation' => 'MySQLd Port',
);

has 'socket' => (
    'is'            => 'ro',
    'isa'           => 'Str',
    'required'      => 0,
    'traits'        => [qw(Getopt)],
    'cmd_aliases'   => 's',
    'documentation' => 'MySQLd Socket',
);

has 'nobinlog' => (
    'is'            => 'ro',
    'isa'           => 'Bool',
    'required'      => 0,
    'default'       => 0,
    'traits'        => [qw(Getopt)],
    'cmd_aliases'   => 'b',
    'documentation' => 'Write command to binlog?',
);

has 'cmd' => (
    'is'            => 'ro',
    'isa'           => 'Str',
    'required'      => 0,
    'traits'        => [qw(Getopt)],
    'cmd_aliases'   => 'c',
    'documentation' => 'A SQL command',
);

has 'file' => (
    'is'            => 'ro',
    'isa'           => 'Str',
    'required'      => 0,
    'traits'        => [qw(Getopt)],
    'cmd_aliases'   => 'f',
    'documentation' => 'A SQL file',
);

has 'host' => (
    'is'            => 'ro',
    'isa'           => 'ArrayRef[Str]',
    'required'      => 1,
    'traits'        => [qw(Getopt)],
    'cmd_aliases'   => 'h',
    'documentation' => 'A list of database hosts to execute all commands on',
);

has '_queries' => (
    'is'      => 'rw',
    'isa'     => 'ArrayRef',
    'default' => sub { [] },
);

# with ...
# initializers ...

# your code here ...
sub _read_sql_file {
    my $self = shift;

    if ( open( my $SQL_FH, '<', $self->file() ) ) {
        my @lines = <$SQL_FH>;

        # DGR: we're just reading ...
        ## no critic (RequireCheckedClose)
        close($SQL_FH);
        ## use critic
        my $tmp_query;
        foreach my $line (@lines) {
            next if $line =~ m/^\s*#/;
            $line =~ s/\s*#.*$//;
            $line =~ s/^\s+//;
            $tmp_query .= $line;
            if ( $tmp_query && $tmp_query =~ m/;\s*$/ ) {
                push( @{ $self->_queries() }, $tmp_query );
            }
            if ( !$self->tables() && $line =~ m/_TABLENAME_/ ) {
                my $msg = 'Need to use --tables switch if you want to use _TABLENAME_ keyword in your queries!';
                $self->logger()->log( message => $msg, level => 'warning', );
                croak($msg);
            }
        } ## end foreach my $line (@lines)
        return 1;
    } ## end if ( open( my $SQL_FH,...))
    return;
} ## end sub _read_sql_file

sub execute {
    my $self = shift;

    if ( !$self->dbs() && !$self->tables() ) {
        my $msg = 'Specify either --tables or --dbs!';
        $self->logger()->log( message => $msg, level => 'warning', );
        croak($msg);
    }
    if ( !$self->cmd() && ( !$self->file() || !-r $self->file() ) ) {
        my $msg = 'Specify --file and point to a valid sql file or use --cmd to define an adhoc query!';
        $self->logger()->log( message => $msg, level => 'warning', );
        croak($msg);
    }

    if ( $self->file() ) {
        $self->_read_sql_file();
    }
    else {
        push( @{ $self->_queries() }, $self->cmd() );
    }

    # add localhost as default if no other host is given!
    if ( scalar( @{ $self->host() } ) < 1 ) {
        push( @{ $self->host() }, 'localhost' );
    }

  HOST: foreach my $host ( @{ $self->host() } ) {
        $self->_process_host($host);
    }    # end foreach

    return 1;
} ## end sub execute

sub _process_host {
    my $self = shift;
    my $host = shift;

    my $DB = MTK::DB::->new(
        {
            'username' => $self->user(),
            'password' => $self->pass(),
            'hostname' => $host,
            'socket'   => $self->socket(),
            'database' => 'mysql',
            'port'     => $self->port(),
            'logger'   => $self->logger(),
        }
    );

    if ( !$DB || !$DB->valid() ) {
        $self->logger()->log( message => 'Connection to '.$host.' failed.', level => 'error', );
        next HOST;
    }
    $self->logger()->log( message => 'Processing Host '.$host.' w/ DSN ' . $DB->dsn(), level => 'notice', );

    # OnlyDB        : STRING    : only return tables from this DB
    # ReturnHashRef : BOOL      : return a hash with extende information instead of the default array of `DB`.`TABLE`?
    # Excludes      : ARRAY     : filter out tables matching these patterns
    # IncludeOnly   : ARRRAY    : only return tables matching these patterns
    # AddHost       : STRING    : add a hostname in front of the table name
    # OnlyEngine    : STRING    : return only tables of this engine type (MyISAM, InnoDB, etc.)
    # NotOnlyBaseTables : BOOL  : return NOT only base tables (i.e. views too)
    my $msg_hdr = 'DB.TABLE - ENGINE - QUERY - STATUS - OUTPUT';
    $self->logger()->log( message => $msg_hdr, level => 'notice', );
    print $msg_hdr. "\n";
    my $table_ref = $DB->list_tables(
        {
            'ReturnHashRef' => 1,
            'Excludes'      => $self->excludes(),
            'Incldues'      => $self->dbs(),
        }
    );
    {

        # enable outflush so the query is printed before the result - so
        # see whats going on. normally its only flushed on each newline.
        ## no critic (ProhibitLocalVars)
        local $OUTPUT_AUTOFLUSH = 1;
        ## use critic
        foreach my $db ( keys %{$table_ref} ) {

            # DGR: 'coz of /i
            ## no critic (ProhibitFixedStringMatches)
            next if $db =~ m/^mysql$/i;
            ## use critic
            foreach my $table ( keys %{ $table_ref->{$db} } ) {
                my $engine = $table_ref->{$db}{$table}{'engine'};
                foreach my $qry (@{$self->queries()}) {
                    my $query = $qry;
                    $query =~ s/_DBNAME_/`$db`/g;
                    $query =~ s/_TABLENAME_/`$table`/g;
                    $DB->do('SET sql_log_bin = 0') if $self->nobinlog();
                    my $stm = $DB->prepare($query);

                    my $msg = "$db.$table - $engine - $query - ";
                    print $msg;

                    if ( $stm->execute() ) {
                        $msg .= 'OK';
                        $self->logger()->log( message => $msg, level => 'notice', );
                        print 'OK - ';
                    }
                    else {
                        $DB->do('SET sql_log_bin = 1') if $self->nobinlog();
                        $msg .= 'ERROR';

                        print "ERROR\n";
                        next;
                    } ## end else [ if ( $stm->execute() )]

                    $self->_print_results($stm);

                    # Done
                    $stm->finish();
                    $DB->do('SET sql_log_bin = 1') if $self->nobinlog();
                } ## end foreach my $qry (@queries)
            } ## end foreach my $table ( keys %{...})
        } ## end foreach my $db ( keys %{$table_ref...})
    }
    $DB->disconnect();

    $self->logger()->log( message => 'Finished Host '.$host, level => 'notice', );
    return 1;
} ## end sub _process_host

sub _print_results {
    my $self = shift;
    my $stm  = shift;

    # fetch in a sensible way, i.e. print column headers and a table
    # of results if there is more than one line/value.
    my $hash_ref = $stm->fetchrow_hashref();

    # the whole idea of this buffer code is to allow a distinction between
    # queries that return only on row and column (like select count(*) from table)
    # and queries that return more than that. these simple queries
    # will get their return value reported on line, the others get a table.
    my $buffer      = q{};
    my $first_col   = undef;
    my $one_row_col = 1;       # only one row in output?
    if ( scalar( keys( %{$hash_ref} ) ) > 1 ) {
        $one_row_col = 0;
    }
    $buffer .= "Result:\n";

    # Heading
    $buffer .= '| ';
    foreach my $key ( keys( %{$hash_ref} ) ) {
        $buffer .= $key.' | ';
    }
    $buffer .= "\n";

    # First row
    $buffer .= '| ';
    foreach my $key ( keys( %{$hash_ref} ) ) {
        if ( !$first_col ) {
            $first_col = $hash_ref->{$key};
        }
        $buffer .= $hash_ref->{$key} . ' | ';
    } ## end foreach my $key ( keys( %{$hash_ref...}))
    $buffer .= "\n";

    # Reamaining rows
    while ( my @row = $stm->fetchrow_array() ) {
        if ($one_row_col) {
            $one_row_col = 0;
            print $buffer;
        }
        print '| ';
        foreach my $row (@row) {
            print $row. ' | ';
        }
        print "\n";
    } ## end while ( my @row = $stm->fetchrow_array...)
    if ($one_row_col) {
        print 'Result: ' . $first_col . "\n";
    }
    return 1;
} ## end sub _print_results

sub abstract {
    return 'Execute a given bit of SQL on every table';
}

no Moose;
__PACKAGE__->meta->make_immutable;

1;

__END__

=head1 NAME

MTK::Cmd::Command::doall - execute a given SQL stmt on any db and/or table

=method abstract

Workaround.

=method execute

Run the doall command.

=head1 ACKNOWLEDGEMENT

This module was originally developed for eGENTIC Systems. With approval from eGENTIC Systems,
this module was generalized and published, for which the authors would like to express their
gratitude.

=cut
